_posts/2020-07-08-event-driven-functions-with-openfaas-and-nats.md

```diff
@@ -0,0 +1,146 @@
+`nats-connector` is connector which invokes OpenFaaS function when a message is sent to a NATS topic.
+
+```
+helm repo add nats-connector  https://openfaas.github.io/faas-netes
+
+helm install nats-connector openfaas/nats-connector --namespace=openfaas
+
+```
+
+> NATS comes with OpenFaaS installation, so we are not setting up NATS here
+
+#### Login to OpenFaaS gateway using CLI
+
+Port forward the gateway service to access it. 
+
+```
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+Get the password
+
+```
+PASSWORD=$(kubectl get secret -n openfaas basic-auth -o jsonpath="{.data.basic-auth-password}" | base64 --decode; echo)
+```
+
+Login to the gateway
+
+```
+echo -n $PASSWORD | faas-cli login --username admin --password-stdin
+```
+
+
+## Deploy functions
+`nats-connector` repo comes with test functions to verify installation. It has two OpenFaas functions.
+
+`receive-message` function subscribes to NATS topic nats-test. On every new message on nats-test topic, receive-message will be invoked.
+
+`publish-message` function publishes message to nats-test topic.
+
+Deploy `receive-message` function
+
+```
+faas-cli deploy --name receive-message --image openfaas/receive-message:latest --fprocess='./handler' --annotation topic="nats-test"
+```
+
+Deploy `publish-message` function
+```
+faas-cli deploy --name publish-message  --image openfaas/publish-message:latest  --fprocess='./handler' --env nats_url=nats://nats.openfaas:4222
+```
+
+> Note: You can also build and deploy this function using the stack.yml and code present in nats-connector repository. But for simplicity, I am using pre built and published images of  these functions.
+
+Invoke `publish-message` function to publish a test message
+
+```
+faas-cli invoke publish-message <<< "test message"
+```
+
+When `publish-message` was invoked, it would have pushed `test-message` to the `nats-test` topic on NATS, which would have invoked `receive-message`. We can verify that by checking logs of `receive-message` function.
+
+```
+faas-cli logs receive-message
+
+ARNING! Communication is not secure, please consider using HTTPS. Letsencrypt.org offers free SSL/TLS certificates.
+Handling connection for 8080
+2020-05-29T15:41:17Z 2020/05/29 15:41:17 Started logging stderr from function.
+2020-05-29T15:41:17Z 2020/05/29 15:41:17 Started logging stdout from function.
+2020-05-29T15:41:17Z Forking - ./handler []
+2020-05-29T15:41:17Z 2020/05/29 15:41:17 OperationalMode: http
+2020-05-29T15:41:17Z 2020/05/29 15:41:17 Timeouts: read: 10s, write: 10s hard: 10s.
+2020-05-29T15:41:17Z 2020/05/29 15:41:17 Listening on port: 8080
+2020-05-29T15:41:17Z 2020/05/29 15:41:17 Writing lock-file to: /tmp/.lock
+2020-05-29T15:41:17Z 2020/05/29 15:41:17 Metrics listening on port: 8081
+2020-05-29T15:42:24Z 2020/05/29 15:42:24 stderr: 2020/05/29 15:42:24 received "test message"
+2020-05-29T15:42:24Z 2020/05/29 15:42:24 POST / - 200 OK - ContentLength: 28
+```
+
+### What Next ?
```

Comment:
@alexellis what do you think about this call to action?

> Do you have a cool "event driven" use case you want to share? Let us know and become a guest blogger!

Reply:
Added